File: api_nestjs_distributed_order_platform/src/application/use-case/order/cancel-order.use-case.ts
import {
    BadRequestException,
    Inject,
    Injectable,
    Logger,
    NotFoundException,
} from "@nestjs/common";
import { InjectConnection } from "@nestjs/sequelize";
import { Sequelize } from "sequelize-typescript";
import type { EventPublisherRepositoryInterface } from "src/application/port/event-publisher.repository";
import type { OrderRepositoryInterface } from "src/application/port/order.repository";
import type { StockRepositoryInterface } from "src/application/port/stock.repository";
import { OrderEntity } from "src/domain/entities/order.entity";
import { StockEntity } from "src/domain/entities/stock.entity";
import { ORDER_REPOSITORY } from "src/shared/tokens_nest/order.token";
import { EVENT_PUBLISHER } from "src/shared/tokens_nest/rabbitmq.token";
import { STOCK_REPOSITORY } from "src/shared/tokens_nest/stock.token";

@Injectable()
export class CancelOrderUseCase {
    private readonly logger = new Logger(CancelOrderUseCase.name);

    constructor(
        @Inject(ORDER_REPOSITORY)
        private readonly orderRepository: OrderRepositoryInterface,
        @Inject(STOCK_REPOSITORY)
        private readonly stockRepository: StockRepositoryInterface,
        @InjectConnection()
        private readonly sequelize: Sequelize,
        @Inject(EVENT_PUBLISHER)
        private readonly eventPublisher: EventPublisherRepositoryInterface,
    ) {}

    async execute(orderId: string): Promise<OrderEntity> {
        const order = await this.orderRepository.findById(orderId);
        if (!order) {
            throw new NotFoundException('Order not found');
        }

        try {
            order.cancelOrder();
        } catch (error) {
            throw new BadRequestException(
                error instanceof Error ? error.message : 'Unable to cancel order',
            );
        }

        const releasedStocks = await this.sequelize.transaction(
            async (transaction): Promise<StockEntity[]> => {
                const stocks: StockEntity[] = [];

                for (const item of order.getItems()) {
                    const stock = await this.stockRepository.release(
                        item.getProductId(),
                        item.getQuantity(),
                        transaction,
                    );
                    stocks.push(stock);
                }

                await this.orderRepository.update(order, transaction);
                return stocks;
            },
        );

        try {
            await this.eventPublisher.publishCancelledOrder(order);
        } catch (error) {
            this.logger.error(
                `Failed to publish OrderCancelled for order ${order.getId()}`,
                error instanceof Error ? error.stack : String(error),
            );
        }

        for (const stock of releasedStocks) {
            try {
                await this.eventPublisher.publishStockReleased(stock);
            } catch (error) {
                this.logger.error(
                    `Failed to publish StockReleased for product ${stock.getProductId()}`,
                    error instanceof Error ? error.stack : String(error),
                );
            }
        }

        return order;
    }
}

File: api_java_distributed_order_platform/src/main/java/com/project/order/application/usecase/order/CancelOrderUseCase.java
package com.project.order.application.usecase.order;

import com.project.order.application.exception.BusinessRuleException;
import com.project.order.application.exception.ConflictException;
import com.project.order.application.exception.NotFoundException;
import com.project.order.application.port.EventPublisherRepository;
import com.project.order.application.port.OrderRepository;
import com.project.order.application.port.StockRepository;
import com.project.order.domain.model.Order;
import com.project.order.domain.model.Stock;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.dao.OptimisticLockingFailureException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class CancelOrderUseCase {

    private static final Logger log = LoggerFactory.getLogger(CancelOrderUseCase.class);

    private final OrderRepository orderRepository;
    private final StockRepository stockRepository;
    private final EventPublisherRepository eventPublisher;

    public CancelOrderUseCase(
            OrderRepository orderRepository,
            StockRepository stockRepository,
            EventPublisherRepository eventPublisher
    ) {
        this.orderRepository = orderRepository;
        this.stockRepository = stockRepository;
        this.eventPublisher = eventPublisher;
    }

    @Transactional
    public Order execute(String orderId) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new NotFoundException("Order not found"));

        try {
            order.cancelOrder();
        } catch (IllegalArgumentException e) {
            throw new BusinessRuleException(e.getMessage());
        }

        for (var item : order.getItems()) {
            Stock stock = stockRepository.findById(item.getProductId())
                    .orElseThrow(() -> new NotFoundException("Stock not found: " + item.getProductId()));
            stock.release(item.getQuantity());
            try {
                stockRepository.update(stock);
            } catch (OptimisticLockingFailureException e) {
                throw new ConflictException("Stock was modified by another request, retry");
            }
            try {
                eventPublisher.publishStockReleased(stock);
            } catch (Exception e) {
                log.error("Failed to publish StockReleased for product {}", stock.getProductId(), e);
            }
        }

        orderRepository.update(order);

        try {
            eventPublisher.publishCancelledOrder(order);
        } catch (Exception e) {
            log.error("Failed to publish OrderCancelled for order {}", order.getId(), e);
        }

        return order;
    }
}
